[Docs] Update Azure Postgres/MySQl guide for Flexi (#20504)
gravitational/teleport: docs/pages/database-access/guides/azure-postgres-mysql.mdx

  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
>
Static configuration of database access for Azure PostgreSQL/MySQL is available

> starting from Teleport `8.1` and Azure database auto-discovery is available

Reviewer: Given that the docs for versions older than v8 are not even available on our docs site anymore, I think it's probably safe to remove the note about 8.1.

Author: Removed.

  Service.
- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Install the Teleport Database Service

(!docs/pages/includes/database-access/token.mdx!)

Install Teleport on the host where you will run the Teleport Database Service:

(!docs/pages/includes/install-linux.mdx!)

Create the Database Service configuration.

<Tabs>
<TabItem label="PostgreSQL">

- Specify the region for your database(s) in `--azure-postgres-discovery`.
- Replace the `--proxy` value with your Teleport proxy address or Teleport cloud
  URI (e.g. `mytenant.teleport.sh:443`):

  ```code
  $ teleport db configure create \
    -o file \
    --proxy=teleport.example.com:443 \
    --token=/tmp/token \
    --azure-postgres-discovery=eastus
  ```

</TabItem>
<TabItem label="MySQL">

- Specify the region for your database(s) in `--azure-mysql-discovery`.
- Replace the `--proxy` value with your Teleport proxy address or Teleport cloud
  URI (e.g. `mytenant.teleport.sh:443`):

  ```code
  $ teleport db configure create \
    -o file \
    --proxy=teleport.example.com:443 \
    --token=/tmp/token \
    --azure-mysql-discovery=eastus
  ```

</TabItem>

> <TabItem label="PostgreSQL and MySQL">

Reviewer 2: ```suggestion
<TabItem label="PostgreSQL and MySQL">

Run the following commands on your Database Service host:
```

Author: ```suggestion
<TabItem label="PostgreSQL and MySQL">

Run the following command on your Database Service host:
```

  ```code
  $ teleport db configure create \
    -o file \
    --proxy=teleport.example.com:443 \
    --token=/tmp/token \
    --azure-mysql-discovery=eastus \
    --azure-postgres-discovery=eastus
  ```

<Admonition type="tip">
This will create two `types` entities in `teleport.yaml`, one for each database
type. This is useful if you want different regions, tags, or labels for each
database type.

Alternatively, you can edit `teleport.yaml` to include both database types in a
single entry:

```yaml
db_service:
  azure:
  - types: ["mysql", "postgres"]
  ...
```

</Admonition>

</TabItem>

</Tabs>

This command will generate a Database Service configuration with Azure MySQL/Postgres
database auto-discovery enabled in the `eastus` region and place it at the
`/etc/teleport.yaml` location.

### Create a Teleport role

On your workstation logged in to your Teleport cluster with `tsh`, define a new

> role to provide access to your Azure database. Our example file is
> `azure-databse-role.yaml`:

Reviewer 2: ```suggestion
role to provide access to your Azure database. Create a file called `azure-database-role.yaml` with the following content:
```

```yaml
version: v6
kind: role
metadata:
  name: azure-database-access
spec:
  allow:
    db_labels:

>       '*': '*'

Reviewer 3: ```suggestion
      'engine':
        - "Microsoft.DBforMySQL/servers"
        - "Microsoft.DBforMySQL/flexibleServers"
        - "Microsoft.DBforPostgreSQL/servers"
        - "Microsoft.DBforPostgreSQL/flexibleServers"
```
nit: I think we should demonstrate how they can restrict the scope of the access role to just Azure mysql/postgres discovered databases. Otherwise it's a little confusing to call the role "azure-database-access" when it grants access to any database

![Teleport Database Access Azure PostgreSQL/MySQL Cloud](../../../img/database-access/guides/azure_cloud.png)
</ScopedBlock>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)
- Deployed Azure Database for PostgreSQL or MySQL server.
- Azure Active Directory administrative privileges.
- A host, e.g., an Azure VM instance, where you will run the Teleport Database
  Service.
- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Install the Teleport Database Service

(!docs/pages/includes/database-access/token.mdx!)

Install Teleport on the host where you will run the Teleport Database Service:

(!docs/pages/includes/install-linux.mdx!)

Create the Database Service configuration.

- Specify the region for your database(s) in `--azure-postgres-discovery`.
- Replace the `--proxy` value with your Teleport proxy address or Teleport cloud

>   URI (e.g. `mytenant.teleport.sh:443`):

Author: Done.

  ```code
  $ teleport db configure create \
    -o file \
    --proxy=teleport.example.com:443 \
    --token=/tmp/token \
    --azure-mysql-discovery=eastus
  ```

</TabItem>

<TabItem label="PostgreSQL and MySQL">

  ```code
  $ teleport db configure create \
    -o file \
    --proxy=teleport.example.com:443 \
    --token=/tmp/token \
    --azure-mysql-discovery=eastus \
    --azure-postgres-discovery=eastus
  ```

<Admonition type="tip">
This will create two `types` entities in `teleport.yaml`, one for each database
type. This is useful if you want different regions, tags, or labels for each
database type.

Alternatively, you can edit `teleport.yaml` to include both database types in a
single entry:

```yaml
db_service:
  azure:
  - types: ["mysql", "postgres"]
  ...
```

</Admonition>

</TabItem>

</Tabs>

This command will generate a Database Service configuration with Azure MySQL/Postgres
database auto-discovery enabled in the `eastus` region and place it at the
`/etc/teleport.yaml` location.

### Create a Teleport role

On your workstation logged in to your Teleport cluster with `tsh`, define a new
role to provide access to your Azure database. Our example file is

> `azure-databse-role.yaml`:

Reviewer: What does "our example file is" mean?

Should we instead say something like, "for example, this role grants access to.."

Author: It means that the example contents below are going into a file with that name. This is relevant for the later step where we create the role by specifying the file name.

Author: Updated.

```yaml

> version: v5

Reviewer: If this is not being backported to v11 or earlier, let's use v6 for the role.

kind: role
metadata:
  name: azure-database-access
spec:
  allow:
    db_labels:
      '*': '*'
    db_names:
    - teleport-mysql
    db_users:

>     - 'teleport'

Reviewer: Why quote teleport here but not teleport-mysql above?

Author: No good reason. Adjusted, and `teleport-mysql` was removed anyway.

```

- The value(s) of `db_names` corresponds to your MySQL database name(s) in

>   Azure, or `postgres` for any PostgreSQL servers.

Reviewer: I'm confused about the postgres comment. Does this mean we only support connecting to Postgres databases named "postgres"?

Reviewer 3: No it's a miscommunication; we support connecting to other db_names within a postgres server. "postgres" is just the default database and what I advised Alex use when he was testing the Flexible server discovery feature himself. 
 The v11 version of this guide had a section, about creating a user, that described these fields accurately and used `db_names`/`db_users` traits with the default `access` role, hence I suggested removing this whole section and using the old section.

Author: I've updated this section with the standard table describing the flags, and used `*` in the example.

- The value of `db_users` corresponds to the database user or role we will create
  later on in the guide.
- Edit `db_labels` to further restrict access for this role.

Save this file and apply it to your Teleport cluster:

```code

> $ tctl create -f azure-databse-role.yaml

Reviewer: ```suggestion
$ tctl create -f azure-database-role.yaml
```

> role 'azure-databse-role.yaml' has been created

Reviewer: ```suggestion
role 'azure-database-role.yaml' has been created
```

> ```

Author: Discussed internally, but to summarize:
- Creating a new user for this guide is against updated conventions discussed in #20024
- Having an RBAC system means creating **R**oles.

@GavinFrazar discussed edits are in if you'd like to comment on additional restrictions.

### Assign Azure AD administrator

Only the Azure AD administrator for the database can connect to it and create
Azure AD users. Go to your database's *Active Directory admin* page
and set the AD admin using the *Set admin* button:
Azure AD users.

<Tabs>
<TabItem label="MySQL">
Go to your database's **Authentication** page and set the AD
admin using the edit button:

![Set AD admin](../../../img/database-access/guides/azure/set-ad-admin.png)

</TabItem>
<TabItem label="PostgreSQL">
Go to your database's **Authentication** page and set the AD
admin by selecting **+ Add Azure AD Admins**:

![Set AD admin](../../../img/database-access/guides/azure/set-ad-admin-postgres.png)

> ![Set AD admin](../../../img/database-access/guides/azure/set-ad-admin@2x.png)

Reviewer 3: We should keep the old steps for setting AD admin with single-server as well, can we add that back as a tab?

Author: Ahh, I assumed this was just a change in the Azure interface, I didn't realize it was specific to flexible servers.

fetch the access token and use it as a password when connecting to the database:

<Tabs>
<TabItem label="PostgreSQL Flexible">

```code
$ az login -u ad@example.com

> $ export PGPASSWORD=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
> $ psql "host=example.postgres.database.azure.com user=ad@example.com sslmode=require dbname=postgres"

Reviewer 3: I originally suggested exporting the token, but let's update this so we don't export it:
```suggestion
$ TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
$ PGPASSWORD=$TOKEN psql "host=example.postgres.database.azure.com user=ad@example.com sslmode=require dbname=postgres"
```

```

</TabItem>
<TabItem label="PostgreSQL">

```code
$ az login -u ad@example.com

> $ export PGPASSWORD=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
> $ psql "host=example.postgres.database.azure.com user=ad@example.com@instance-name sslmode=require dbname=postgres"

Reviewer 3: ```suggestion
$ TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
$ PGPASSWORD=$TOKEN psql "host=example.postgres.database.azure.com user=ad@example.com@instance-name sslmode=require dbname=postgres"
```
same as above

```

</TabItem>
<TabItem label="MySQL Flexible">

```code
$ az login -u ad-admin@example.com

> $ export TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`

Reviewer 3: ```suggestion
$ TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
```
same idea as above

$ mysql -h example.mysql.database.azure.com -P 3306 -u ad@example.com --enable-cleartext-plugin --password=$TOKEN
```

</TabItem>
<TabItem label="MySQL">

```code
$ az login -u ad-admin@example.com

> $ export TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`

Reviewer 3: ```suggestion
$ TOKEN=`az account get-access-token --resource-type oss-rdbms --output tsv --query accessToken`
```
same as above
